Fix incorrect workaround for pure Y (X) rotations in XYZ (YXZ) conven…

tagcup opened the pull request:
…tion.

Fixes #10352.

tagcup committed:
Fix incorrect workaround for pure Y (X) rotations in XYZ (YXZ) convention.

Fixes #10352.

akien-mga commented on core/math/matrix3.cpp:
```diff
@@ -436,15 +432,10 @@ Vector3 Basis::get_euler_yxz() const {
-			if (elements[1][0] == 0 && elements[0][1] == 0 && elements[2][2] < 0) { // use pure x rotation
-				real_t x = asin(-m12);
+			if (elements[1][0] == 0 && elements[0][1] == 0 && elements[0][2] == 0 && elements[2][0] == 0 && elements[0][0] == 1) { // use pure x rotation
+				real_t x = atan2(-m12, elements[1][1]);
```
`x` is unused since you removed the chunk below, I guess it should be `euler.x`?

tagcup replied:
Indeed, thanks!

tagcup commented:
This of course doesn't fix the problem regarding the non-uniqueness of polar decomposition, which makes it impossible to change scaling without changing rotation part, vice versa.

The behavior of Spatial::set_scale was changed in #7438, where I introduced polar decomposition in Basis::get_scale and Basis::get_rotation (because earlier behavior wasn't mathematically correct in general). So Spatial::set_scale in 2.x and 3.x does different things.

I'll remove (or at least unexport to GDScript) Basis::set_scale and Basis::set_rotation, which I introduced in #8277 in a later PR, and their get_ counterparts need some thinking as well.

reduz commented:
oh, nice, fixes makes a lot more sense than what I did

On Mon, Aug 21, 2017 at 7:54 PM, Rémi Verschelde <notifications@github.com>
wrote:

> Merged #10415 <https://github.com/godotengine/godot/pull/10415>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/10415#event-1214960759>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z21w0RPeU1iIbgmraeAdtwLcsutQvks5sagq7gaJpZM4O7Hkc>
> .
>

tagcup commented:
Clarification for others: this simply checks whether we have a [simple rotation around the Y (X) axis](https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations) (there should be 4 zeroes and a single 1), and proceeds to extract the rotation angle.
